File: electron/fileManager.ts
import * as fs from 'node:fs';
import * as fsp from 'node:fs/promises';
import * as path from 'node:path';
import * as os from 'node:os';

/** Application configuration */
export interface AppConfig {
  vrmModelPath: string | null;
  animationFolderPath: string | null;
  windowPosition: { x: number; y: number };
  windowSize: { width: number; height: number };
  scale: number;
  micEnabled: boolean;
  cameraEnabled: boolean;
  targetFps: number;
  powerSaveMode: boolean;
  autonomousMovementPaused: boolean;
  animationLoopEnabled: boolean;
  autoExpressionEnabled: boolean;
  allowedAutoExpressions: string[];
  animationSpeed: number;
  moveSpeedMultiplier: number;
  systemAssetsDir: string;
  /** VRM 模型瀏覽對話框上次使用的資料夾（為空時從 vrmModelPath 推導） */
  vrmPickerFolder?: string;
  mtoonOutlineEnabled?: boolean;
  agent?: AgentConfig;
  headTracking?: HeadTrackingConfig;
}

/** 滑鼠頭部追蹤設定（與 src/types/config.ts HeadTrackingConfig 同步） */
export interface HeadTrackingConfig {
  enabled: boolean;
  weight: number;
  smoothingRate: number;
}

/**
 * my-agent 整合設定（與 src/types/config.ts AgentConfig 同步）。
 *
 * v0.4 之前：subprocess + ws（spawn `cli daemon` + AgentDaemonManager / AgentSessionClient）
 * v0.4 起（M-MASCOT-EMBED Phase 5）：in-process via AgentRuntime + vendor/my-agent/dist-embedded
 *
 * 舊欄位（daemonMode / bunBinaryPath / myAgentCliPath）保留以利平滑遷移；
 * AgentRuntime 不讀，預期 v0.5 完成後可刪除。
 */
export interface AgentConfig {
  /** Master toggle — 啟用 my-agent 功能（會 preload LLM 進入 standby） */
  enabled: boolean;
  /** Workspace 目錄；預設 ~/.virtual-assistant-desktop/agent-workspace */
  workspaceCwd: string | null;
  /** 本地 LLM 設定 */
  llm: {
    /** GGUF 模型絕對路徑；null 時 toggle ON 會 error */
    modelPath: string | null;
    /** Context window size；預設 4096 */
    contextSize: number;
    /** GPU layers；'auto' 由 node-llama-tcq 決定 */
    gpuLayers: number | 'auto';
    /** 替代方案：外部 llama.cpp HTTP endpoint（embedded 模式以外）*/
    externalUrl: string | null;
  };
  /** Opt-in daemon WS server（讓外部 my-agent CLI / 第二個 Electron window 連） */
  daemon: {
    enabled: boolean;
    port: number;
  };
  /** Opt-in web UI HTTP server */
  webUi: {
    enabled: boolean;
    port: number;
  };
  /** Legacy 欄位（v0.3.x agent，subprocess mode）— AgentRuntime 不讀，預期 v0.5 移除 */
  daemonMode?: 'auto' | 'external';
  bunBinaryPath?: string | null;
  myAgentCliPath?: string | null;
}

/** Animation entry metadata */
export interface AnimationEntry {
  fileName: string;
  displayName: string;
  category: string;
  loop: boolean;
  weight: number;
}

/** Animation metadata collection */
export interface AnimationMeta {
  folderPath: string;
  entries: AnimationEntry[];
}

const DEFAULT_CONFIG: AppConfig = {
  vrmModelPath: null,
  animationFolderPath: null,
  windowPosition: { x: 0, y: 0 },
  windowSize: { width: 400, height: 600 },
  scale: 1.0,
  micEnabled: false,
  cameraEnabled: false,
  targetFps: 30,
  powerSaveMode: false,
  autonomousMovementPaused: false,
  animationLoopEnabled: true,
  autoExpressionEnabled: true,
  allowedAutoExpressions: [],
  animationSpeed: 1.0,
  moveSpeedMultiplier: 1.0,
  systemAssetsDir: 'assets/system',
  mtoonOutlineEnabled: false,
  agent: {
    enabled: false,
    workspaceCwd: null,
    llm: {
      modelPath: null,
      contextSize: 4096,
      gpuLayers: 'auto',
      externalUrl: null,
    },
    daemon: {
      enabled: false,
      port: 0,
    },
    webUi: {
      enabled: false,
      port: 0,
    },
    // legacy 欄位保留（舊 config 反向相容；新版 AgentRuntime 不讀）
    daemonMode: 'auto',
    bunBinaryPath: null,
    myAgentCliPath: null,
  },
  headTracking: {
    enabled: true,
    weight: 0.7,
    smoothingRate: 4,
  },
};

/** Get config directory path (~/.virtual-assistant-desktop/) */
export function getConfigDir(): string {
  return path.join(os.homedir(), '.virtual-assistant-desktop');
}

function getConfigPath(): string {
  return path.join(getConfigDir(), 'config.json');
}

function getAnimationMetaPath(): string {
  return path.join(getConfigDir(), 'animations.json');
}

/** Ensure config directory exists */
export async function ensureConfigDir(): Promise<void> {
  const dir = getConfigDir();
  await fsp.mkdir(dir, { recursive: true });
}

/** Check if config.json exists */
export function getConfigExists(): boolean {
  return fs.existsSync(getConfigPath());
}

/** Read config.json, auto-backup and recreate if corrupted */
export async function readConfig(): Promise<AppConfig> {
  const configPath = getConfigPath();

  try {
    await fsp.access(configPath);
  } catch {
    return { ...DEFAULT_CONFIG };
  }

  try {
    const content = await fsp.readFile(configPath, 'utf-8');
    const parsed = JSON.parse(content) as Partial<AppConfig>;
    // 深合併 agent 內巢狀欄位（llm / daemon / webUi），舊 schema 的 config 自動補齊缺欄位
    const defaultAgent = DEFAULT_CONFIG.agent!;
    const parsedAgent: Partial<AgentConfig> = parsed.agent ?? {};
    const mergedAgent: AgentConfig = {
      ...defaultAgent,
      ...parsedAgent,
      llm: { ...defaultAgent.llm, ...(parsedAgent.llm ?? {}) },
      daemon: { ...defaultAgent.daemon, ...(parsedAgent.daemon ?? {}) },
      webUi: { ...defaultAgent.webUi, ...(parsedAgent.webUi ?? {}) },
    };
    return {
      ...DEFAULT_CONFIG,
      ...parsed,
      agent: mergedAgent,
      headTracking: { ...DEFAULT_CONFIG.headTracking!, ...(parsed.headTracking ?? {}) },
    };
  } catch (e) {
    console.warn(`[FileManager] config.json corrupted: ${e}. Backing up and recreating.`);
    const backupPath = path.join(getConfigDir(), 'config.json.bak');
    try {
      await fsp.copyFile(configPath, backupPath);
    } catch {
      // Backup might fail if original is unreadable
    }
    const defaultConfig = { ...DEFAULT_CONFIG };
    await writeConfig(defaultConfig);
    return defaultConfig;
  }
}

/** Write config.json */
export async function writeConfig(config: AppConfig): Promise<void> {
  await ensureConfigDir();
  const content = JSON.stringify(config, null, 2);
  await fsp.writeFile(getConfigPath(), content, 'utf-8');
}

/** Read animations.json */
export async function readAnimationMeta(): Promise<AnimationMeta> {
  const metaPath = getAnimationMetaPath();

  try {
    await fsp.access(metaPath);
  } catch {
    return { folderPath: '', entries: [] };
  }

  try {
    const content = await fsp.readFile(metaPath, 'utf-8');
    return JSON.parse(content) as AnimationMeta;
  } catch (e) {
    console.warn(`[FileManager] animations.json parse error: ${e}`);
    return { folderPath: '', entries: [] };
  }
}

/** Write animations.json */
export async function writeAnimationMeta(meta: AnimationMeta): Promise<void> {
  await ensureConfigDir();
  const content = JSON.stringify(meta, null, 2);
  await fsp.writeFile(getAnimationMetaPath(), content, 'utf-8');
}

/** Scan .vrm files in specified folder, returns full paths */
export async function scanVrmFiles(folderPath: string): Promise<string[]> {
  try {
    const stat = await fsp.stat(folderPath);
    if (!stat.isDirectory()) return [];
  } catch {
    return [];
  }

  const entries = await fsp.readdir(folderPath);
  return entries
    .filter((f) => f.toLowerCase().endsWith('.vrm'))
    .sort()
    .map((f) => path.join(folderPath, f));
}

/** Scan .vrma files in specified folder */
export async function scanVrmaFiles(folderPath: string): Promise<string[]> {
  try {
    const stat = await fsp.stat(folderPath);
    if (!stat.isDirectory()) {
      throw new Error(`Animation folder does not exist: ${folderPath}`);
    }
  } catch (e) {
    throw new Error(`Animation folder does not exist: ${folderPath} (${e})`);
  }

  const entries = await fsp.readdir(folderPath);
  return entries
    .filter((f) => f.toLowerCase().endsWith('.vrma'))
    .sort();
}

/** Sync scanned results with existing metadata */
export function syncAnimationMeta(
  folderPath: string,
  existing: AnimationMeta,
  scannedFiles: string[],
): AnimationMeta {
  const entries: AnimationEntry[] = [];

  for (const fileName of scannedFiles) {
    const existingEntry = existing.entries.find((e) => e.fileName === fileName);
    if (existingEntry) {
      entries.push(existingEntry);
    } else {
      const displayName = fileName.replace(/\.vrma$/i, '');
      entries.push({
        fileName,
        displayName,
        category: 'action',
        loop: false,
        weight: 1.0,
      });
    }
  }

  return { folderPath, entries };
}

/** Scan animations folder and sync with metadata */
export async function scanAnimations(folderPath: string): Promise<AnimationMeta> {
  const scannedFiles = await scanVrmaFiles(folderPath);
  const existing = await readAnimationMeta();
  const synced = syncAnimationMeta(folderPath, existing, scannedFiles);
  await writeAnimationMeta(synced);
  return synced;
}

File: src-tauri/src/file_manager.rs
use serde::{Deserialize, Serialize};
use std::fs;
use std::path::PathBuf;

/// 應用程式設定
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AppConfig {
    pub vrm_model_path: Option<String>,
    pub animation_folder_path: Option<String>,
    pub window_position: Position,
    pub window_size: Size,
    pub scale: f64,
    pub mic_enabled: bool,
    pub camera_enabled: bool,
    pub target_fps: u32,
    pub power_save_mode: bool,
    pub autonomous_movement_paused: bool,
    #[serde(default = "default_true")]
    pub animation_loop_enabled: bool,
    #[serde(default = "default_true")]
    pub auto_expression_enabled: bool,
    #[serde(default)]
    pub allowed_auto_expressions: Vec<String>,
}

fn default_true() -> bool {
    true
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Position {
    pub x: f64,
    pub y: f64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Size {
    pub width: f64,
    pub height: f64,
}

/// 動畫條目
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AnimationEntry {
    pub file_name: String,
    pub display_name: String,
    pub category: String,
    pub r#loop: bool,
    pub weight: f64,
}

/// 動畫 metadata
#[derive(Debug, Clone, Default, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AnimationMeta {
    pub folder_path: String,
    pub entries: Vec<AnimationEntry>,
}

impl Default for AppConfig {
    fn default() -> Self {
        Self {
            vrm_model_path: None,
            animation_folder_path: None,
            window_position: Position { x: 0.0, y: 0.0 },
            window_size: Size {
                width: 400.0,
                height: 600.0,
            },
            scale: 1.0,
            mic_enabled: false,
            camera_enabled: false,
            target_fps: 30,
            power_save_mode: false,
            autonomous_movement_paused: false,
            animation_loop_enabled: true,
            auto_expression_enabled: true,
            allowed_auto_expressions: Vec::new(),
        }
    }
}

/// 取得設定目錄路徑 (~/.virtual-assistant-desktop/)
pub fn get_config_dir() -> Result<PathBuf, String> {
    dirs::home_dir()
        .map(|home| home.join(".virtual-assistant-desktop"))
        .ok_or_else(|| "Cannot determine home directory".to_string())
}

/// 取得 config.json 路徑
fn get_config_path() -> Result<PathBuf, String> {
    Ok(get_config_dir()?.join("config.json"))
}

/// 取得 animations.json 路徑
fn get_animation_meta_path() -> Result<PathBuf, String> {
    Ok(get_config_dir()?.join("animations.json"))
}

/// 讀取 config.json，損毀時自動備份並重建
pub fn read_config_file() -> Result<AppConfig, String> {
    let path = get_config_path()?;

    if !path.exists() {
        return Ok(AppConfig::default());
    }

    let content = fs::read_to_string(&path)
        .map_err(|e| format!("Failed to read config.json: {}", e))?;

    match serde_json::from_str::<AppConfig>(&content) {
        Ok(config) => Ok(config),
        Err(e) => {
            // 設定檔損毀：備份並重建
            log::warn!("config.json is corrupted: {}. Backing up and recreating.", e);
            let backup_path = get_config_dir()?.join("config.json.bak");
            let _ = fs::copy(&path, &backup_path);

            let default_config = AppConfig::default();
            write_config_file(&default_config)?;
            Ok(default_config)
        }
    }
}

/// 寫入 config.json
pub fn write_config_file(config: &AppConfig) -> Result<(), String> {
    let path = get_config_path()?;
    let dir = get_config_dir()?;

    if !dir.exists() {
        fs::create_dir_all(&dir)
            .map_err(|e| format!("Failed to create config dir: {}", e))?;
    }

    let content = serde_json::to_string_pretty(config)
        .map_err(|e| format!("Failed to serialize config: {}", e))?;

    fs::write(&path, content)
        .map_err(|e| format!("Failed to write config.json: {}", e))
}

/// 讀取 animations.json
pub fn read_animation_meta_file() -> Result<AnimationMeta, String> {
    let path = get_animation_meta_path()?;

    if !path.exists() {
        return Ok(AnimationMeta::default());
    }

    let content = fs::read_to_string(&path)
        .map_err(|e| format!("Failed to read animations.json: {}", e))?;

    serde_json::from_str(&content)
        .map_err(|e| format!("Failed to parse animations.json: {}", e))
}

/// 寫入 animations.json
pub fn write_animation_meta_file(meta: &AnimationMeta) -> Result<(), String> {
    let path = get_animation_meta_path()?;
    let dir = get_config_dir()?;

    if !dir.exists() {
        fs::create_dir_all(&dir)
            .map_err(|e| format!("Failed to create config dir: {}", e))?;
    }

    let content = serde_json::to_string_pretty(meta)
        .map_err(|e| format!("Failed to serialize animations.json: {}", e))?;

    fs::write(&path, content)
        .map_err(|e| format!("Failed to write animations.json: {}", e))
}

/// 掃描指定資料夾內的 .vrma 檔案
pub fn scan_vrma_files(folder_path: &str) -> Result<Vec<String>, String> {
    let path = PathBuf::from(folder_path);

    if !path.exists() || !path.is_dir() {
        return Err(format!("Animation folder does not exist: {}", folder_path));
    }

    let mut files = Vec::new();

    let entries = fs::read_dir(&path)
        .map_err(|e| format!("Failed to read animation folder: {}", e))?;

    for entry in entries {
        let entry = entry
            .map_err(|e| format!("Failed to read directory entry: {}", e))?;
        let file_path = entry.path();

        if let Some(ext) = file_path.extension() {
            if ext.eq_ignore_ascii_case("vrma") {
                if let Some(name) = file_path.file_name() {
                    files.push(name.to_string_lossy().to_string());
                }
            }
        }
    }

    files.sort();
    Ok(files)
}

/// 同步掃描結果與現有 metadata
///
/// - 新發現的 .vrma 預設歸類為 action
/// - 已不存在的條目移除
pub fn sync_animation_meta(
    folder_path: &str,
    existing: &AnimationMeta,
    scanned_files: &[String],
) -> AnimationMeta {
    let mut entries = Vec::new();

    for file_name in scanned_files {
        // 嘗試沿用既有設定
        if let Some(existing_entry) = existing.entries.iter().find(|e| &e.file_name == file_name) {
            entries.push(existing_entry.clone());
        } else {
            // 新檔案，預設歸類為 action
            let display_name = file_name
                .strip_suffix(".vrma")
                .or_else(|| file_name.strip_suffix(".VRMA"))
                .unwrap_or(file_name)
                .to_string();

            entries.push(AnimationEntry {
                file_name: file_name.clone(),
                display_name,
                category: "action".to_string(),
                r#loop: false,
                weight: 1.0,
            });
        }
    }

    AnimationMeta {
        folder_path: folder_path.to_string(),
        entries,
    }
}
